src/storage/mvcc/reader/scanner.rs

```diff
@@ -97,34 +99,408 @@ impl<S: Snapshot> BackwardScannerBuilder<S> {
-        let lock_cursor = CursorBuilder::new(&self.snapshot, CF_LOCK)
+    /// Build `ForwardScanner` from the current configuration.
+    pub fn build(self) -> Result<Scanner<S>> {
+        let lock_cursor_builder = CursorBuilder::new(&self.snapshot, CF_LOCK)
             .range(self.lower_bound.clone(), self.upper_bound.clone())
-            .fill_cache(self.fill_cache)
-            .scan_mode(ScanMode::Backward)
-            .build()?;
+            .fill_cache(self.fill_cache);
 
-        let write_cursor = CursorBuilder::new(&self.snapshot, CF_WRITE)
+        let write_cursor_builder = CursorBuilder::new(&self.snapshot, CF_WRITE)
             .range(self.lower_bound.clone(), self.upper_bound.clone())
-            .fill_cache(self.fill_cache)
-            .scan_mode(ScanMode::Backward)
-            .build()?;
+            .fill_cache(self.fill_cache);
+        if self.desc {
+            let lock_cursor = lock_cursor_builder.scan_mode(ScanMode::Backward).build()?;
+            let write_cursor = write_cursor_builder.scan_mode(ScanMode::Backward).build()?;
+            Ok(Scanner::Backward(BackwardScanner::new(
+                self,
+                lock_cursor,
+                write_cursor,
+            )))
+        } else {
+            let write_cursor = write_cursor_builder.build()?;
+            let lock_cursor = lock_cursor_builder.build()?;
+            Ok(Scanner::Forward(ForwardScanner::new(
+                self,
+                lock_cursor,
+                write_cursor,
+            )))
+        }
+    }
+}
+
+/// This struct can be used to scan keys starting from the given user key (greater than or equal).
+///
+/// Internally, for each key, rollbacks are ignored and smaller version will be tried. If the
+/// isolation level is SI, locks will be checked first.
+///
+/// Use `ScannerBuilder` to build `ForwardScanner`.
+pub struct ForwardScanner<S: Snapshot> {
+    snapshot: S,
+    fill_cache: bool,
+    omit_value: bool,
+    isolation_level: IsolationLevel,
+
+    /// `lower_bound` and `upper_bound` is used to create `default_cursor`. `lower_bound`
+    /// is used in initial seek as well. They will be consumed after `default_cursor` is being
+    /// created.
+    lower_bound: Option<Key>,
+    upper_bound: Option<Key>,
+
+    ts: u64,
+
+    lock_cursor: Cursor<S::Iter>,
+    write_cursor: Cursor<S::Iter>,
+
+    /// `default cursor` is lazy created only when it's needed.
+    default_cursor: Option<Cursor<S::Iter>>,
+
+    /// Is iteration started
+    is_started: bool,
+
+    statistics: Statistics,
+}
 
-        Ok(BackwardScanner {
-            snapshot: self.snapshot,
-            fill_cache: self.fill_cache,
-            omit_value: self.omit_value,
-            isolation_level: self.isolation_level,
-            lower_bound: self.lower_bound,
-            upper_bound: self.upper_bound,
-            ts: self.ts,
+impl<S: Snapshot> ForwardScanner<S> {
+    pub fn new(
+        builder: ScannerBuilder<S>,
```

Comment:
It looks wired to accept a builder here. Logically it doesn't make sense for an instance creation function itself accepts an instance builder. What you need may be actually a kind of options.